### PR TITLE
[Enhancement] Support complex type expression push down

### PR DESCRIPTION
**Pull request by Seaven**
Fixes #issue

1. Support complex type expression push down
2. Rename ConjunctivePredicatesRewriter to GlobalDictPredicatesRewriter

## What type of PR is this:
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Checklist:
- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.1
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 5 / 6 (83.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/SubfieldExpr.java|0|1|00.00%|[97]|
|:large_blue_circle:|com/starrocks/analysis/CollectionElementExpr.java|5|5|100.00%|[]|

**Review by stdpain — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/analysis/SubfieldExpr.java [hunk 1]
```diff
@@ -91,4 +91,9 @@ protected void toThrift(TExprNode msg) {
     public Expr clone() {
         return new SubfieldExpr(this);
     }
+
+    @Override
+    public boolean isSelfMonotonic() {
+        return children.get(0).isSelfMonotonic();
```
Comment (fzhedu) [inline fe/fe-core/src/main/java/com/starrocks/analysis/SubfieldExpr.java:97]: why only the first? not all children?

File: be/src/storage/rowset/array_column_iterator.cpp [hunk 2]
```diff
@@ -20,15 +20,17 @@
 
 namespace starrocks {
 
-ArrayColumnIterator::ArrayColumnIterator(std::unique_ptr<ColumnIterator> null_iterator,
+ArrayColumnIterator::ArrayColumnIterator(ColumnReader* reader, std::unique_ptr<ColumnIterator> null_iterator,
```
Comment (fzhedu) [inline be/src/storage/rowset/array_column_iterator.cpp:23]: const?

File: be/src/exec/olap_scan_prepare.cpp [hunk 3]
```diff
@@ -750,18 +750,16 @@ void OlapScanConjunctsManager::build_column_expr_predicates() {
         // otherwise we don't need this limitation.
         const SlotDescriptor* slot_desc = slots[index];
         LogicalType ltype = slot_desc->type().type;
-        if (!is_scalar_logical_type(ltype)) continue;
-        // disable on float/double type because min/max value may lose precision
-        // The fix should be on storage layer, and this is just a temporary fix.
-        if (ltype == LogicalType::TYPE_FLOAT || ltype == LogicalType::TYPE_DOUBLE) continue;
```
Comment (fzhedu) [inline be/src/exec/olap_scan_prepare.cpp:756]: why remove this?

File: fe/fe-core/src/main/java/com/starrocks/analysis/SubfieldExpr.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/analysis/SubfieldExpr.java:97]: second child must a string

File: be/src/exec/olap_scan_prepare.cpp (hunk 3 above)
Comment (Seaven) [reply be/src/exec/olap_scan_prepare.cpp:756]: I add a new method to resolved this, so don't need check again

File: be/src/storage/rowset/array_column_iterator.cpp (hunk 2 above)
Comment (Seaven) [reply be/src/storage/rowset/array_column_iterator.cpp:23]: it's will assign to class's private parameter, i think don't need const?

File: be/src/storage/rowset/map_column_iterator.h [hunk 4]
```diff
@@ -41,14 +42,13 @@ class MapColumnIterator final : public ColumnIterator {
 
     /// for vectorized engine
     Status get_row_ranges_by_zone_map(const std::vector<const ColumnPredicate*>& predicates,
-                                      const ColumnPredicate* del_predicate, SparseRange* row_ranges) override {
-        CHECK(false) << "array column does not has zone map index";
-        return Status::OK();
-    }
```
Comment (silverbullet233) [inline be/src/storage/rowset/map_column_iterator.h:47]: why change here? map/struct/array still doesn't support zone_map filter., right?

Comment (Seaven) [reply be/src/storage/rowset/map_column_iterator.h:47]: yeah, but will support later

**Review by silverbullet233 — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.1

**Pull request merged**